### PR TITLE
ghc: 9.0.2

### PR DESCRIPTION
**Pull request by fishtreesugar**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by fishtreesugar** [comment]
@SMillerDev @carlocab @chenrui333 pls add `CI-long-timeout` label

**Comment by XuehaiPan** [comment]
I think the build dependency `python@3.9` can be changed to `python@3.10`.

**Comment by fishtreesugar** [comment]
@XuehaiPan I'll try after `CI-long-timeout` label added

**Comment by fishtreesugar** [comment]
@SMillerDev @carlocab @chenrui333 happy new year ping :D

**Comment by XuehaiPan** [comment]
The build error indicates the linker cannot find `gmp`.

File: Formula/ghc.rb [hunk 1]
```diff
@@ -1,70 +1,43 @@
...
-      sha256 "fd4829912cddd12f84181c3451cc752be224643e87fac497b69edddadc49b4f2"
-    end
-  end
-
-  on_linux do
-    depends_on "gmp" => :build
-  end
-
-  # https://www.haskell.org/ghc/download_ghc_8_10_7.html#macosx_x86_64
+  # https://www.haskell.org/ghc/download_ghc_9_0_2.html#macosx_x86_64
```
Comment (XuehaiPan) [inline Formula/ghc.rb:24]:
```suggestion
  on_macos do
    resource "gmp" do
      url "https://ftp.gnu.org/gnu/gmp/gmp-6.2.1.tar.xz"
      mirror "https://gmplib.org/download/gmp/gmp-6.2.1.tar.xz"
      mirror "https://ftpmirror.gnu.org/gmp/gmp-6.2.1.tar.xz"
      sha256 "fd4829912cddd12f84181c3451cc752be224643e87fac497b69edddadc49b4f2"
    end
  end

  on_linux do
    depends_on "gmp" => :build
  end

  # https://www.haskell.org/ghc/download_ghc_9_0_2.html#macosx_x86_64
```

File: Formula/ghc.rb [hunk 2]
```diff
@@ -73,44 +59,19 @@ def install
     ENV["LD"] = "ld"
     ENV["PYTHON"] = Formula["python@3.9"].opt_bin/"python3"
```
Comment (XuehaiPan) [inline Formula/ghc.rb:60]: This needs to be changed to `python@3.10` too.

File: Formula/ghc@8.10.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    on_linux do
+      url "https://downloads.haskell.org/~ghc/8.10.7/ghc-8.10.7-x86_64-deb9-linux.tar.xz"
+      sha256 "ced9870ea351af64fb48274b81a664cdb6a9266775f1598a79cbb6fdd5770a23"
+    end
+  end
+
+  def install
+    ENV["CC"] = ENV.cc
+    ENV["LD"] = "ld"
+    ENV["PYTHON"] = Formula["python@3.9"].opt_bin/"python3"
```
Comment (XuehaiPan) [inline Formula/ghc@8.10.rb:67]: This needs to be changed to `python@3.10` too.

```suggestion
    ENV["PYTHON"] = Formula["python@3.10"].opt_bin/"python3"
```

Comment (XuehaiPan) [reply Formula/ghc@8.10.rb:67]: Same changes for `Formula/ghc@9.2.rb`.

**Comment by fishtreesugar** [comment]
I'll take another look after hls support ghc 9.0.2 https://github.com/haskell/haskell-language-server/issues/2541

**Comment by fishtreesugar** [comment]
https://github.com/Homebrew/homebrew-core/pull/94049 yay

File: Formula/ghc.rb [hunk 4]
```diff
@@ -1,33 +1,19 @@
...
   homepage "https://haskell.org/ghc/"
-  url "https://downloads.haskell.org/~ghc/8.10.7/ghc-8.10.7-src.tar.xz"
-  sha256 "e3eef6229ce9908dfe1ea41436befb0455fefb1932559e860ad4c606b0d03c9d"
-  # We bundle a static GMP so GHC inherits GMP's license
-  license all_of: [
-    "BSD-3-Clause",
-    any_of: ["LGPL-3.0-or-later", "GPL-2.0-or-later"],
-  ]
-  revision 1
+  url "https://downloads.haskell.org/~ghc/9.0.2/ghc-9.0.2-src.tar.xz"
```
Comment (carlocab) [inline Formula/ghc.rb:4]: Is it time to do this already? `ghcup`'s recommended version is still 8.10, so I thought that's still the version that upstream recommends.

File: Formula/ghc.rb [hunk 5]
```diff
@@ -73,44 +59,19 @@ def install
...
-                              "--build=#{cpu}-apple-darwin#{OS.kernel_version.major}"
-        system "make"
-        system "make", "install"
-      end
-
-      args = ["--with-gmp-includes=#{gmp}/include",
-              "--with-gmp-libraries=#{gmp}/lib"]
-    end
+    # https://gitlab.haskell.org/ghc/ghc/-/issues/20592#note_391266
+    ENV["C_INCLUDE_PATH"] = `xcrun --show-sdk-path`.strip + "/usr/include/ffi" if Hardware::CPU.arm? && OS.mac?
```
Comment (carlocab) [inline Formula/ghc.rb:63]:
```suggestion
    ENV["C_INCLUDE_PATH"] = MacOS.sdk_path_if_needed/"usr/include/ffi" if Hardware::CPU.arm? && OS.mac?
```

**Review by carlocab — COMMENTED** [review]
Looks like upstream binaries still link with `gmp`. If so, maybe we want to keep the GMP dependency?

File: Formula/ghc.rb (hunk 4 above)
Comment (fishtreesugar) [reply Formula/ghc.rb:4]: Good question. I don't know the criteria of ghcup treat a GHC version as recommended.
IMO, GHC 9.0.2 is a bug fix release, supported by ghcup, hls and 9.0.2 [stackage nightly](https://www.stackage.org/nightly-2022-01-27) support more package than GHC 8.10.7 [LTS](https://www.stackage.org/lts-18.23), which means GHC 9.0.2 mature enough for more adoption.
Another benefit of GHC 9.0.2 is, GHC 9.x has better compiling performance then 8.10, it *may* reduce some  CI resource of Haskell projects

**Comment by fishtreesugar** [comment]
I close it for waiting LTS of GHC 9.0.2

**Comment by carlocab** [comment]
Thanks @fishtreesugar. Do you know when that might be?

Next time, it would be good to split this PR into multiple separate ones to try to minimise the CI time required.

**Comment by fishtreesugar** [comment]
@carlocab I dunno. 

> Next time, it would be good to split this PR into multiple separate ones to try to minimise the CI time required.

IC. I'll try my best 🙏🏻

**Comment by fishtreesugar** [comment]
@carlocab  [Stackage LTS of GHC 9.0.2](https://www.stackage.org/lts-19.0) already release, want to know how to reduce CI time by split this PR.

**Comment by carlocab** [comment]
Let's start with a single PR that renames `ghc@9` to `ghc@9.2`. You'll need to add an alias `ghc@9` that points to `ghc@9.2`.

We can do `9.0` and `8.10` separately when that is merged.

**Pull request closed without merging**